java/src/main/java/net/snowflake/floe/FloeDecryptorImpl.java

```diff
@@ -20,90 +21,58 @@ public byte[] processSegment(byte[] input) {
   }
 
   @Override
-  public byte[] processSegment(byte[] input, int offset, int length) {
+  public byte[] processSegment(byte[] input, int offset, final int length) {
+    if (length == -1) {
+      return processSegment(input, offset, 0);
+    }
     return processInternal(() -> {
-      ByteBuffer inputBuffer = ByteBuffer.wrap(input, offset, length);
+      ByteBuffer inputBuf = ByteBuffer.wrap(input, offset, length);
       try {
-        if (isLastSegment(inputBuffer)) {
-          return processLastSegment(inputBuffer);
-        } else {
-          return processNonLastSegment(inputBuffer);
+        verifyMinimalSegmentLength(inputBuf);
+        verifySegmentNotTooLong(inputBuf);
+        boolean isTerminal = isTerminal(inputBuf);
+        verifySegmentSizeWithSegmentSizeMarker(inputBuf, isTerminal);
+        AeadKey aeadKey = getKey(messageKey, floeIv, floeAad, segmentCounter);
+        AeadIv aeadIv = AeadIv.from(inputBuf, parameterSpec.getAead().getIvLength());
+        AeadAad aeadAad = isTerminal ? AeadAad.terminal(segmentCounter) : AeadAad.nonTerminal(segmentCounter);
+        byte[] decrypted =
+            aeadProvider.decrypt(aeadKey, aeadIv, aeadAad, inputBuf.array(), inputBuf.position(), inputBuf.remaining());
```

Comment:
NitPick: I know that we wrap the array above so that we know this code is safe, but I prefer to write it more conservatively and demonstrate proper use of `ByteBuffer` objects to readers.
```suggestion
            aeadProvider.decrypt(aeadKey, aeadIv, aeadAad, inputBuf.array(), inputBuf.position() + inputBuf.arrayOffset(), inputBuf.remaining());
```